### PR TITLE
[Blog]: adding a blog for rl time-slicing

### PR DESCRIPTION
**Pull request by poonam-lamba**
## What does this PR do?

New blog for time-slicing feature

## Why is this change needed?

Time-slicing is launched, blog is needed to make folks aware of the change

## How was this tested?
All of the below

- [ ] Tests added/updated (`npm test`)
- [x] Site builds successfully (`npm run build:all`)
- [x] Check links after buildling (`npm run check-links`)
- [x] Manual testing performed (`npm run serve`)

## Checklist

- [ ] Commits are signed off (`git commit -s`) per [DCO](../PR_SIGNOFF.md)
- [ ] Code follows project [contributing guidelines](../CONTRIBUTING.md)
- [ ] Tests pass locally (`npm test`)
- [ ] Site builds without errors (`npm run build:all`)
- [ ] No broken links after building full site (`npm run check-links`)
- [ ] Documentation updated (if applicable)

## Related Issues

[](https://github.com/llm-d/llm-d/issues/1668)

### CODE REVIEW

**Comment by davidgs** [comment]
Please resolve the conflicts and sign your commits.

**Review by chcost — COMMENTED** [review]
Thanks for putting this together, @poonam-lamba  time-slicing is a genuinely exciting addition to the project and this blog does a good job explaining the architecture and the end-to-end flow.

A few suggestions to bring it in line with the style and tone of the other llm-d blogs (heterogeneous inference, BLIS, networking, hybrid models, etc.):

Tone a neutrality: The existing blogs are project-centric. So in general the subject is always "llm-d" or a named component, never "we" in a way that implies a single organization. A few passages here read more like a product announcement than a community contribution (e.g., "We foresee this trend early on", "llm-d is not a reactive patch"). I'd suggest reframing these to let the technical content make the case, which it does well on its own.

Structure: The existing blogs tend to open with a tight 1-3 sentence problem statement and get to the feature quickly. Here, the first three paragraphs and the "Core Challenges" section cover overlapping ground before we reach time-slicing. Tightening the intro and cutting the repetition would help. Consider adding a "Key Takeaways" box at the top (like some of the other blog do).

Benchmark rigor: The other blogs are specific about their benchmark setups (GPU count, model, workload config, methodology caveats). The "Early Benchmarks" section here would benefit from those details, plus a limitations/caveats section — BLIS, heterogeneous inference, and networking blogs all include one.

Connection to llm-d architecture: the blog introduces several new components (Snapshot Agent, Accelerator Orchestrator, TimeSlice Client) but doesn't position them relative to the existing llm-d architecture (router/EPP, model servers, KV cache hierarchy). The other blogs are careful to show where new capabilities plug into the existing stack. Worth considering adding some wording making the connection.

Happy to discuss any of these in more detail. Specific suggestions in inline comments.

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+authors:
+  - poonamlamba
+  - bogdan
+  - aishu
+  - dolev
+tags: [blog, updates, llm-d, rl]
+---
+
+# RL Post-Training: Co-Operative Time-Slicing
+
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing-with-llm-d.md:15]: This opening is strong on motivation but could be tightened. The existing blogs open with 1-3 sentences framing the problem, then move directly to the solution. Consider cutting this down to a single paragraph and moving straight to time-slicing by the second paragraph.

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+The math behind Reinforcement Learning (RL) post-training for Large Language Models is notoriously unforgiving. As frontier AI labs push the boundaries of reasoning and coding models using algorithms like GRPO, they routinely hit hard architectural and physical constraints. While much of the industry's focus remains on raw GPU count, the actual gatekeeper of experimental velocity is infrastructure efficiency.
+
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing-with-llm-d.md:17]: This might read as a specific organization claiming foresight rather than a community project announcement. We are typically carefully and avoid this pattern (e.g. from other blogs "BLIS provides a faster inner loop") or the networking blog ("NIXL presents a single API"). Suggest a slight reframing.

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+As established in large-scale systems research including ByteDance’s [HybridFlow](https://arxiv.org/abs/2409.19256) architecture (powering veRL) and Alibaba’s [RollMux](https://arxiv.org/abs/2512.11306), optimizing the ratio of generator (sampler) to trainer throughput is the single largest driver of RL Total Cost of Ownership (TCO). While these frameworks define the architecture, executing this at production scale is challenging. We foresee this trend early on and have been purposefully building llm-d to address efficiency bottlenecks in RL. llm-d is not a reactive patch, it is a mature, production tested engine that is being deployed across production RL workloads.
+
+We have made targeted investments to expand llm-d into a comprehensive, composable RL infrastructure stack designed to meet researchers where they are i.e. whether they are using Slurm or Kubernetes based setups:
+
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing-with-llm-d.md:21]: A couple of concerns with this section:

1. Agent Sandbox (Sister Project) — we should clarify what is the relationship to llm-d? The other blogs don't reference related projects without clear context.
2. Our Agent Sandbox provides... — "Our" is possessive in a way the other blogs avoid. Suggest: "The Agent Sandbox provides..."
3. More broadly, this section tries to cover the entire RL stack in three bullet points before the blog has introduced its actual subject (time-slicing). Consider either expanding this into its own blog post or trimming it to a single sentence that links to a tracking issue.

On the RL scheduler specifically: The blog introduces the [py-inference-scheduler](https://github.com/llm-d-incubation/py-inference-scheduler) as doing "intelligent routing and batching mechanics for the sampling step" — but this is also what the llm-d Router (EPP) does for inference workloads. The relationship between these two components isn't explained, and a reader familiar with llm-d's architecture would reasonably ask: why is there a second scheduler?

The relationship between py-scheduler and the EPP for RL workloads is still evolving and is active work within sig-rl. I think we have two options:

1. Acknowledge it briefly.. e.g., "The RL scheduler complements the llm-d Router for RL-specific batch orchestration patterns (acquire/yield semantics, weight-version-aware dispatching) that are still being integrated into the core routing layer. This integration is an active area of work across sig-rl and sig-router."
2. Omit the routing details entirely and keep this blog focused purely on time-slicing, which is its strongest content. The RL scheduler and weight propagation interface could be covered in their own blog posts when the EPP integration story is clearer.

I'd lean toward option 2, the time-slicing content is strong enough to stand on its own, and bundling three other half-introduced components dilutes the message.

File: blog/2026-06-28_rl_time-slicing.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+We have made targeted investments to expand llm-d into a comprehensive, composable RL infrastructure stack designed to meet researchers where they are i.e. whether they are using Slurm or Kubernetes based setups:
+
+1. **Throughput-Driven Inference**: The llm-d router maximises the RL rollout generation throughput, ensuring samplers can saturate the pipeline and prevent downstream optimization trainers from stalling.
+2. **Agent Sandbox (Sister Project)**: Fast tool use and isolated code execution are vital for generating reward signals. Our Agent Sandbox provides the secure, sub-second execution environments necessary to scale verifiable rewards and agentic rollouts without bottlenecking the GPUs.
+3. **Core Pipeline Primitives**: To combat generation stragglers, we shipped asynchronous batching and native inference schedulers, specifically designed to orchestrate complex, multi-turn RL workloads.
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing.md]: Suggest: "Today, llm-d introduces a new well-lit path for Co-operative Time-Slicing: the Snapshot Agent."

File: blog/2026-06-28_rl_time-slicing.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+Today, we are introducing a new well-lit path in the **llm-d** project for Co-operative Time-Slicing : Snapshot Agent. Time-Slicing effort enables job interleaving into the Kubernetes platform layer to address the single greatest source of waste in modern RL: the Idle Accelerators. Snapshot agent is the first component that we are launching, other components will soon follow.
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing.md]: "The following two sections in the blog provide a high level overview..." — this kind of meta-narration ("in this blog we will...") is not used in the other posts. The reader can follow the structure from the headings. Suggest removing this paragraph.

File: blog/2026-06-28_rl_time-slicing.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+The following two sections in the blog provide a high level overview of problems in running RL workloads on scale followed by a brief overview of investments for solving those problems and then we dive deeper into the time-slicing for RL.
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing.md]: This section largely restates what was already covered in the opening paragraphs. The existing blogs state the problem once, then move to the solution. Suggest merging this with a trimmed intro so the problem is stated once and concisely.

File: blog/2026-06-28_rl_time-slicing.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+## Core Challenges in Enterprise RL Infrastructure
+
+Deploying large-scale RL infrastructure presents operational friction and cost inefficiencies. Main inefficiency among these is low hardware utilization; because typical RL training loops suffer from alternating blocking phases leading to an average GPU/TPU duty cycle of only around 40%. This structural waste is exacerbated during the sampling step, where generation stragglers create significant tail latencies and drag down overall sampling efficiency.
+
+Furthermore, scaling distributed RL remains an engineering bottleneck. Teams face difficulties in configuring network primitives like NCCL and Intra-Cluster Interconnect (ICI), while simultaneously battling slow checkpointing operations and high-latency weight transfers across decoupled pools.
+
+Finally, as organizations pivot toward Agentic RL use cases, they need isolation for untrusted code execution and tool calling during generation and reward evaluation steps during RL, all while struggling with reliability when running RL in distributed environments at scale.
+
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing.md]: Minor: space before the colon is non-standard. Also, "llm-d have been investing" should be "llm-d has been investing."

File: blog/2026-06-28_rl_time-slicing.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+## The Solution : A composable RL Infrastructure Stack
+
+To resolve these bottlenecks, llm-d have been investing in improving the ***efficiency, security, and observability*** for RL workloads. The solutions are composable meaning you can adopt one or more depending on your use case.
+
+**Efficiency:**
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing.md]: The Weight Propagation Interface and RL scheduler get one-line descriptions with repo links but no explanation of how they integrate with the existing llm-d architecture. The other blogs are careful to show how new components plug into the stack (e.g., the networking blog shows exactly where NixlConnector sits in the vLLM → NIXL → network path).

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+We believe researchers should focus on core modeling logic rather than wrestling with low-level CUDA context switching or custom scheduling loops. With llm-d, you can enable RL job multiplexing by wrapping the accelerator phases in your existing loops with a simple Python decorator:
+
+```python
+from timeslice import OrchestratorClient
+
+orchestrator = OrchestratorClient(job_id="job-1")
+
+@orchestrator.on_accelerators(group_id="trainer-group")
+def train_phase(model, trajectories):
+    return model.update(trajectories)
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing-with-llm-d.md:124]: The other blogs specify exact benchmark configurations. Could you add:
- How many H100 GPUs?
- What model and size?
- What workload / batch configuration?
- How was the 41% baseline measured?

Also, "theoretical peak of ~95% under idealized phase alignments" could use a brief derivation or citation. Consider adding a limitations/caveats section — the heterogeneous inference blog has a "workload caveat" note, BLIS has a full "Limitations" section. E.g., what assumptions does the 95% number depend on? What workload characteristics would reduce the benefit?

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+**Framework integrations**: We will provide well-lit paths for integrating with slime, veRL, and other popular RL training frameworks making it easy for researchers to adopt time-slicing without leaving their existing training stack.
+
+**Simplified onboarding and intelligent job placement**: We will simplify platform deployment and, looking further ahead, build a component that automatically identifies time-sliceable workloads and makes intelligent job placement decisions eliminating the need for users to manually identify and group compatible workloads.
+
+**Expanding accelerator support**: In line with llm-d's hardware-agnostic mission, we are designing the platform to span multiple accelerator architectures starting with GPU and expanding to TPU and beyond.
+
+### Help Shape the Future of SIG-RL
+
+Building robust, highly optimized RL infrastructure requires tight collaboration with the engineers and researchers running these workloads at scale.
+
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing-with-llm-d.md:152]: Broken link — missing closing parenthesis. Should be `...main)`.

File: blog/authors.yml [hunk 11]
```diff
@@ -268,3 +268,28 @@ abhisyant:
   image_url: /img/blogs/abhisyant.webp
   socials:
     linkedin: abhisyant
+
+bogdan:
```
Comment (chcost) [inline blog/authors.yml:292]: `bogdan` and `dolev` are missing `image_url`, which is inconsistent with other author entries. Consider adding GitHub avatar URLs for consistency.

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md (hunk 1 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing-with-llm-d.md:15]: updated

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md (hunk 2 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing-with-llm-d.md:17]: I removed this framing, ptal

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md (hunk 3 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing-with-llm-d.md:21]: Great feedback, this blog is now just focused on time-slicing. I removed the references to scheduler and sandbox

File: blog/2026-06-28_rl_time-slicing.md (hunk 4 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing.md]: updated.

File: blog/2026-06-28_rl_time-slicing.md (hunk 6 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing.md]: addressed

File: blog/2026-06-28_rl_time-slicing.md (hunk 8 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing.md]: removed references to WPI and scheduler

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md (hunk 10 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing-with-llm-d.md:152]: updated

File: blog/authors.yml (hunk 11 above)
Comment (poonam-lamba) [reply blog/authors.yml:292]: fixed

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md (hunk 9 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing-with-llm-d.md:124]: I added more explanation, however if it is still ambiguous then we can remove it altogether until we have a benchmark which demonstrates this. For now, we have a baseline of 40% and we are able to push it to 70%, we believe there is more room beyond this based on various jobs that were alayzed.

File: blog/2026-06-28_rl_time-slicing.md (hunk 5 above)
Comment (poonam-lamba) [reply blog/2026-06-28_rl_time-slicing.md]: I updated the section based on feedback

**Comment by chcost** [comment]
Not sure why we are making changes to the no-K8s blog date in this PR. Can you please remove that file from the PR?

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  - bogdan
+  - aishu
+  - dolev
+tags: [blog, updates, llm-d, rl]
+---
+
+# RL Post-Training: Co-Operative Time-Slicing
+
+In Reinforcement Learning (RL) post-training for Large Language Models using algorithms like GRPO, optimizing the ratio of generator (sampler) to trainer throughput is the single largest driver of infrastructure Total Cost of Ownership (TCO). Because typical RL post-training loops alternate sequentially between generation and optimization phases, expensive GPU and TPU clusters sit completely idle for 40% to 60% of their lifecycle.
+
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing-with-llm-d.md:17]: In llm-d, "well-lit path" refers to the fully supported, end-to-end deployment patterns (intelligent scheduling, P/D disaggregation, wide expert-parallelism) that ship with production guides, CI validation, and release artifacts. Time-slicing isn't there yet. It's an incubation-stage component with a single guide. Suggest using "integration guide" for now. As the RL work matures and meets the bar for CI coverage, docs, and release support, it can graduate to a well-lit path.

Suggested rewording:

> Today, we introduce an integration guide for Co-operative Time-Slicing with the Snapshot Agent.

File: blog/2026-06-28_rl_time-slicing-with-llm-d.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+- **TimeSlice Orchestrator and TimeSlice Client**: The TimeSlice Orchestrator coordinates time-slicing across multiple jobs, managing co-operative lock queues and driving context switches at phase boundaries. Alongside it, a lightweight TimeSlice Client library will allow any RL application to integrate using a two-call API: `acquire()` before the accelerator phase and `yield()` after.
+- **Framework Integrations**: Future well-lit paths will integrate with slime, veRL, and other popular RL training frameworks, making it easy for researchers to adopt time-slicing without leaving their existing training stack.
+- **Simplified Onboarding and Intelligent Job Placement**: Ongoing work will simplify platform deployment and introduce a component that automatically identifies time-sliceable workloads and makes intelligent job placement decisions, eliminating the need for users to manually identify and group compatible workloads.
+- **Expanding Accelerator Support**: In line with llm-d's hardware-agnostic mission, the platform is designed to span multiple accelerator architectures, starting with GPUs and expanding to TPUs and beyond.
+
```
Comment (chcost) [inline blog/2026-06-28_rl_time-slicing-with-llm-d.md:148]: Same note as above on "well-lit path." Suggested rewording:

> Today, we introduce an integration guide and the Snapshot Agent for integrating Co-operative Time-Slicing into managed multi-tenant RL services to support full fine-tuning.

**Review by chcost — COMMENTED** [review]
Two small wording fixes on "well-lit path" usage.

**Comment by davidgs** [comment]
@poonam-lamba Please go back and sign all of your commits as soon as you can.

**Review by davidgs — APPROVED** [review]
lgtm
